src/client/Microsoft.Identity.Client/Instance/Region/RegionManager.cs

```diff
@@ -151,18 +154,10 @@ private bool IsTelemetryRecorded(ApiEvent apiEvent)
 
         private async Task<RegionInfo> DiscoverAndCacheAsync(string azureRegionConfig, ICoreLogger logger, CancellationToken requestCancellationToken)
         {
-            if (s_failedAutoDiscovery == true)
-            {
-                var autoDiscoveryError = $"[Region discovery] Auto-discovery failed in the past. Not trying again. {s_regionDiscoveryDetails}. {DateTime.UtcNow}";
-                logger.Verbose(autoDiscoveryError);
-                return new RegionInfo(null, RegionAutodetectionSource.FailedAutoDiscovery, autoDiscoveryError);
-            }
-
-            if (s_failedAutoDiscovery == false &&
-                !string.IsNullOrEmpty(s_autoDiscoveredRegion))
-            {
-                logger.Info($"[Region discovery] Auto-discovery already ran and found {s_autoDiscoveredRegion}.");
-                return new RegionInfo(s_autoDiscoveredRegion, RegionAutodetectionSource.Cache, null);
+            var regionInfo = GetCachedRegion(logger);
+            if (regionInfo != null)
+            {
+                return regionInfo;
             }
 
             var result = await DiscoverAsync(logger, requestCancellationToken).ConfigureAwait(false);
```

Comment:
These need to be set inside of a lock. Otherwise
- Thread 1 discovers region, releases semaphore.
- Thread 2 calls GetCachedRegion and uses old values, sees that region was not found, starts discovery.
- Thread 1 updates the static region values to the region it found.

```csharp
            s_failedAutoDiscovery = result.RegionSource == RegionAutodetectionSource.FailedAutoDiscovery;
            s_autoDiscoveredRegion = result.Region;
            s_regionDiscoveryDetails = result.RegionDetails;
```

Reply:
Good catch